samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/BeginDeliveryActionListener.java

```diff
@@ -453,10 +455,22 @@ else if (minute > 1) {
     		if (extTimeService.hasExtendedTime() && extTimeService.getTimeLimit() > 0) {
     			control.setTimeLimit(extTimeService.getTimeLimit());
     		}
-    		String timeLimitInSetting = control.getTimeLimit() == null ? "0" : control.getTimeLimit().toString();
     		Date attemptDate = unSubmittedAssessmentGrading.getAttemptDate();
+    		long timeLimitInSetting = control.getTimeLimit();
+    		Long now = new Date().getTime();
+    		Long start = attemptDate.getTime();
+    		if((now - start) > (timeLimitInSetting*1000)) {
+    			// check if the queue is ahead and already submitted it
+    			TimedAssessmentQueue queue = TimedAssessmentQueue.getInstance();
+    			TimedAssessmentGradingModel timedAG = (TimedAssessmentGradingModel)queue.
+    					get(unSubmittedAssessmentGrading.getAssessmentGradingId());
+    			// if it was submitted (race condition) while checking, unblock it - sam will synch soon
+    			if(timedAG != null && !timedAG.getSubmittedForGrade()) {
+    				delivery.setTimeExpired(true);
+    			}
+    		}
     		delivery.setBeginTime(attemptDate);
-    		String timeBeforeDueRetract = delivery.getTimeBeforeDueRetract(timeLimitInSetting);
+    		String timeBeforeDueRetract = delivery.getTimeBeforeDueRetract(control.getTimeLimit() == null ? "0" : String.valueOf(control.getTimeLimit()));
```

Comment:
Personally, I'm of the opinion that we should utilize highly accepted and tested code from library's such as Apache Commons, rather than "re-inventing the wheel" so to speak. It also helps to improve code readability and maintainability.

I'm open to other's opinions though.